### PR TITLE
Remove old UNSAFE_ lifecycle methods

### PR DESCRIPTION
**Pull request by mfix22**
~I am not sure if this is a valid fix yet, but I was going to let CI run the tests for me. I'll close and look into it if the build fails.~

Let me know if this will cause issues, but I don't think it should. The React docs recommends moving `componentWillMount` logic into the constructor

### CODE REVIEW

**Comment by timneutkens** [comment]
Are head update reflected in time before rendering when doing this though? I remember @giuseppeg saying it works fine for styled-jsx, so maybe he can weight in 🕵️ 

Thank you very much for this contribution, helps a lot 😌

**Comment by mfix22** [comment]
@timneutkens I believe so, since constructor is called before `UNSAFE_componentWillMount`, but I am also not an expert on Next.js (:

**Comment by giuseppeg** [comment]
If side effects are synchronous it should be good :)

https://github.com/zeit/styled-jsx/pull/457#issuecomment-400329970

File: lib/side-effect.js [hunk 1]
```diff
@@ -77,9 +76,6 @@ export default function withSideEffect (reduceComponentsToState, handleStateChan
       }
     }
 
-    // Make UNSAFE_ compatible with version of React under 16.3
-    polyfill(SideEffect)
```
Comment (mfix22) [inline lib/side-effect.js]: @timneutkens I see that a couple tests still use `UNSAFE_` . . . do you want me to keep this line?

Comment (timneutkens) [reply lib/side-effect.js]: Tests don't need the compat because we run tests on 16.4

Comment (timneutkens) [reply lib/side-effect.js]: So removing it is perfect 👍

**Comment by timneutkens** [comment]
I just realized we don't have integration tests for routing with next/head, could you add a few? Otherwise, I'll add them 👍

**Comment by mfix22** [comment]
@timneutkens I'm not sure if I have time right now so you will probably get it done a lot sooner than me, but I'll try and get to it 👍

**Comment by mfix22** [comment]
Actually @timneutkens I'm not _exactly_ sure what you mean by: 
> routing with next/head

It would be helpful to have some clarification on what exactly to test for, and if I should create a new folder or add it to an existing one.

**Comment by timneutkens** [comment]
You can extend `test/integration/basic` there's navigation tests already, what I meant is that we don't have any tests doing client-side routing that changes the `<head>` using `next/head`. So the way to go is creating 2 new pages, and then navigate between them and check if the `<head>` is updated in both instances 🕵️

**Comment by timneutkens** [comment]
Thank you very much for adding the test, I've updated it a bit, this is ready to be merged after tests pass 😌

**Review by timneutkens — APPROVED** [review]

**Comment by mfix22** [comment]
Thanks Tim 👍

**Comment by SBoudrias** [comment]
Any chance this broke `Head` inside `dynamic()` components? I'm seeing this doesn't work anymore as of `7-canary.3`

**Comment by timneutkens** [comment]
@SBoudrias could you create a minimal reproduction so that we can add a test for it?

**Pull request merged**